src/ghga_connector/core/downloading/batch_processing.py

```diff
@@ -129,167 +129,108 @@ def handle_response(self, *, response: str):
+        if not is_service_healthy(dcs_api_url):
+            raise exceptions.ApiNotReachableError(api_url=dcs_api_url)
+        self.api_url = dcs_api_url
+        self.message_display = message_display
+        self.work_package_accessor = work_package_accessor
         self.max_wait_time = max_wait_time
-        # amount of seconds between staging attempts
-        self.retry_after = retry_after
-
-    def check_api_available(self):
-        """Get response from endpoint, else throw corresponding exception"""
-        if not is_service_healthy(self.api_url):
-            raise exceptions.ApiNotReachableError(api_url=self.api_url)
-
-    def get_file_ids(self) -> list[str]:
-        """Get file ids that should be staged"""
-        return list(self.file_ids_with_extension.keys())
-
-    def remove_existing(self, *, file_ids: list[str]):
-        """Delete file information for already existing files."""
-        for file_id in file_ids:
-            del self.file_ids_with_extension[file_id]
-
-
-@dataclass
-class StagingState:
-    """Handle state for staged and not yet staged file ids."""
-
-    time_started: datetime = field(default_factory=datetime.utcnow, init=False)
-    staged_files: list[str] = field(default_factory=list, init=False)
-    unstaged_files: list[str] = field(default_factory=list, init=False)
-
-    def add_staged(self, *, file_id: str):
-        """Delegate adding staged file ids"""
-        self.staged_files.append(file_id)
-
-    def add_unstaged(self, *, file_id: str):
-        """Delegate adding unstaged file ids"""
-        self.unstaged_files.append(file_id)
-
-    def update_staged_files_wait(
-        self, *, max_wait_time: int, work_package_accessor: WorkPackageAccessor
-    ) -> bool:
-        """
-        Update staged file list after all previously staged files have been processed.
-        Caller has to make sure, that all file ids in self.staged_files have actually
-        been processed.
-
-        Returns a boolean instructing the caller to wait for retry_time and call this
-        method again.
-        """
-        self.staged_files = []
-        remaining_unstaged = []
-
-        for file_id in self.unstaged_files:
+        self.time_started = now = time()
+        self.staged = {}
+        self.unstaged = {
+            file_id: now
+            for file_id in wanted_file_ids
+            if file_id not in existing_file_ids
+        }
+        self.failed = []
+        self.ignore_failed = False
+
+    def get_staged_files(self):
+        """Get files that are already staged."""
+        staging_items = list(self.unstaged.items())
+        for file_id, retry_time in staging_items:
+            if time() >= retry_time:
+                self._check_file(file_id=file_id)
+        if not self.staged and not self._handle_failures():
+            sleep(1)
+        self._check_timeout()
+        staged = self.staged.copy()
+        self.staged.clear()
+        return staged
```

Comment:
A brief comment explaining the reason `self.staged` is being copied and cleared instead of used directly would be nice

Reply:
b7ded70 moves this responsibility to the caller and mentions it in the docstring.

Reply:
Looks great, thank you.